cleangoldenami/main.go

```diff
@@ -58,23 +60,60 @@ func main() {
 				log.Fatalf("error fetching images for region %v: %v", regionName, err)
 			}
 
-			if imageCount := len(images); imageCount < quota {
+			if imageCount := len(images); (quota - imageCount) >= desiredImageCapacity {
 				if *verbose {
 					fmt.Printf("Region %v is under quota. (Images: %v, Quota: %v)\n", regionName, imageCount, quota)
 				}
 			} else {
-				imageToDelete, err := getOldestImage(images)
-				if err != nil {
-					log.Fatalf("error determining which image to delete in region %v: %v", regionName, err)
-				}
-				if *dryRun {
-					fmt.Printf("Region %v is at quota (%v) - would delete %v\n", regionName, quota, *imageToDelete.ImageId)
-				} else {
-					err := deregisterImage(ec2Client, imageToDelete)
+				numOfImagesToDelete := desiredImageCapacity - (quota - imageCount)
+
+				arm64Images := filterImages(images, "arm64", "rhel-arm64")
+				legacyx86Images := filterImages(images, "x86_64", "legacy-x86_64")
+				x86Images := filterImages(images, "x86_64", "rhel-x86_64")
+
+				var imagesToDelete []ec2Types.Image
+
+				for !imageDeletionCheck(imagesToDelete, numOfImagesToDelete, arm64Images, legacyx86Images, x86Images) {
+					// Returns the most populated image type and the slice containing images of that type
+					mostPopulatedImagesByType, mostPopulatedImageType := getMostPopulatedImageType(arm64Images, legacyx86Images, x86Images)
```

Comment:
I'm not familiar enough with Go slices and how they behave as pointers (or not) off the top of my head, but I'm wondering if we could _only_ return the slice here, and then down below instead of the switch block have something like:
```
if ImageToDeleteIndex >= 0 {
	// Add the image to the list of images to delete
	imagesToDelete = append(imagesToDelete, ImageToDelete)
	mostPopulatedImagesByType = append(mostPopulatedImagesByType[:ImageToDeleteIndex], mostPopulatedImagesByType[ImageToDeleteIndex+1:]...)
}
```

Reply:
Did some testing on my own here - I don't think this will work sadly.

Reply:
yeah I tested it and came to the same conclusion